cardano-api/src/Cardano/Api/Typed.hs

```diff
@@ -1647,88 +1650,96 @@ toJSONmOfnChecks keys required total
                       ++ " number of key hashes: " ++ show (length keys)
 
 instance FromJSON MultiSigScript where
-  parseJSON = Aeson.withObject "MultiSigScript" $ \obj -> return (parseMultiSigScript obj)
+  parseJSON = Aeson.withObject "MultiSigScript" $ \obj -> parseMultiSigScript obj
 
 
-parseMultiSigScript :: HMS.HashMap Text Value -> MultiSigScript
-parseMultiSigScript hms = case  all' hms <|> any' hms <|> atLeast hms of
-                            Just mss -> mss
-                            Nothing -> error "No multi-signature script found"
+parseMultiSigScript :: HMS.HashMap Text Value -> Aeson.Parser MultiSigScript
+parseMultiSigScript hms = do anyM <- any' hms
+                             allM <- all' hms
+                             atLeastM <- atLeast hms
+                             case anyM <|> allM <|> atLeastM of
```

Comment:
Do we need to do it this way, with several parsers returning `Maybe`? The `aeson` parser is an instance of `Alternative` itself. So wouldn't it work to use `<|>` directly on the parsers for any/all/atLeast, and then those individual parsers can use the ordinary style where they expect exactly what they're after, rather than having to be in an optional style.